src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs

```diff
@@ -0,0 +1,162 @@
+    {
+        this.configuredTimeoutInMs = (long)timeout.TotalMilliseconds;
+    }
+
+    public event EventHandler<RequestTelemetryState> TelemetryStateTimedOut;
+
+    public void PushTelemetryState(Message request, RequestTelemetryState telemetryState)
+    {
+        var messageId = request?.Headers.MessageId?.ToString();
+        if (messageId != null)
+        {
+            var expiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + this.configuredTimeoutInMs;
+            var entry = new Entry(telemetryState, expiresAt);
+            lock (this.sync)
+            {
+                this.outstandingRequestStates.Add(messageId, entry);
+                if (this.timer == null)
+                {
+                    this.SetTimer(expiresAt);
+                }
+            }
+        }
+    }
+
+    public RequestTelemetryState PopTelemetryState(Message reply)
+    {
+        var relatesTo = reply?.Headers.RelatesTo?.ToString();
+        return relatesTo == null ? null : this.PopTelemetryState(relatesTo);
+    }
+
+    public void Dispose()
+    {
+        this.SetTimer(null);
+
+        Entry[] abandonedEntries = null;
+        lock (this.sync)
+        {
+            abandonedEntries = new Entry[this.outstandingRequestStates.Count];
+            this.outstandingRequestStates.Values.CopyTo(abandonedEntries, 0);
+            this.outstandingRequestStates.Clear();
+        }
+
+        foreach (Entry entry in abandonedEntries)
+        {
+            this.TriggerTimedOutEvent(entry);
+        }
+    }
+
+    private RequestTelemetryState PopTelemetryState(string messageId)
+    {
+        lock (this.sync)
+        {
+            var entry = (Entry)this.outstandingRequestStates[messageId];
+            if (entry != null)
+            {
+                this.outstandingRequestStates.Remove(messageId);
+            }
+
+            return entry?.State;
+        }
+    }
+
+    private void OnTimer(object state)
+    {
+        List<DictionaryEntry> timedOut = null;
+        lock (this.sync)
+        {
+            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            long? nextExpiry = null;
+            foreach (DictionaryEntry kvp in this.outstandingRequestStates)
+            {
+                if (((Entry)kvp.Value).ExpiresAt <= now)
+                {
+                    if (timedOut == null)
+                    {
+                        timedOut = new List<DictionaryEntry>();
+                    }
+
+                    timedOut.Add(kvp);
+                }
+                else
+                {
+                    nextExpiry = ((Entry)kvp.Value).ExpiresAt;
+                    break;
+                }
+            }
+
+            timedOut?.ForEach(kvp => this.outstandingRequestStates.Remove(kvp.Key));
+            this.SetTimer(nextExpiry);
+        }
+
+        timedOut?.ForEach(kvp => this.TriggerTimedOutEvent((Entry)kvp.Value));
+    }
+
+    private void SetTimer(long? dueAt)
```

Comment:
Instead of using `Nullable<long>` and having to check if it has a value or not and then retrieve that value, you could use `Timeout.Infinite`. When telling `System.Threading.Timer` that the timer is due in `Timeout.infinite`, it effectively turns off the timer. Then instead of disposing the existing timer and creating a new one, you could use `timer.Change(dueAt, Timeout.Infinite)` to modify the existing timer to fire at the new due time. You will need to wrap in a try/catch as `Timer.Change` throws `ObjectDisposedException` if there's a race and you dispose the Timer in the trackers Dispose method. Doing things this way would mean you would need to call `Timer.Dispose` in your Dispose method.